### DISCUSSION TITLE
Comparing more than 2 datasets & comparing values in a single table

### DISCUSSION THREAD

**Discussion by poorbadger** [moj-analytical-services/splink#2312] [Q&A]
Hello All! 

Questions on two use cases....

What's best practice when comparing across more than 2 datasets? 

Say you've got 10 datasets (or 100) and you're comparing products with a manufacturer, product name, upc which is often absent and a unit price.

Would you just compare dataset 1 to 2, combine them, compare new combined dataset to dataset 3, combine, compare to dataset 4 and on and on?

What if you have duplicates within a single table and you're attempting to deduplicate/consolidate that table?

Thank you!

**Comment by dcaseykc** [comment]
When I've done this, I've stacked all my datasets together and treat the problem as one big de-duplication. I'm not sure if this violates any statistical assumptions or is otherwise naughty, but it seems to work in practice.

**Comment by medwar99** [comment]
I take the approach of doing separate internal linkage and clustering per dataset (Splink calls this a dedupe-only, but I don't actually drop any rows). I then perform one big link-only of all the sets together, obtaining only the pairwise predictions (not clusters for this step). I then perform a union-all of all these identified edges together (i.e. _n_ intra-set pairwise edge tables + 1 inter-set pairwise edge table). I then perform one last cluster on that union to get the cross-set cluster IDs.

I could do a dedupe-and-link model, but I feel this process works better for my workflow, especially when I have fields specific to a given dataset which can produce strong intra-set links but might not be present in the other sets. It also helps when I have complex blocking rules that are easier to write within the scope of a single set. You also achieve per-set deduplication as a nice part of the process, and I can keep the sets separate until I want to do something with them, joining on the cluster_id.

Chaining linkages as in your example (1->2->3...) could produce bottlenecks through which you may see degraded performance. A multi-set linkage can at least get around that by producing cross-set edges without needing to chain through a specific set.

**Comment by RobinL** [comment]
Thanks for the answers!  I think all the suggestions are legitimate and it depends a bit on the use case.  Internally, our process is very close to what @medwar99 describes.

I think broadly there's a complexity/accuracy tradeoff here.  

The 'stack everything together' solution is simplest.  But you're essentially averaging a model across many datasets, meaning fewer parameters, so you're potentially leaving accuracy on the table.

One important consideration is the variety in the 10 datasets.

- If they have the same columns and similar data quality, then a single model is probably fine - you'd probably not get much more accuracy with multiple models
 
- If the datasets have a common set of columns, but each some have additional fields specific to the dataset, then the 'within dataset' models have the potential to be more accurate than the 'across dataset' models, because the 'within dataset' models have access to the additional fields.  e.g. dataset 1 may have cols A, B, C and dataset 2 has only two columns A, B.  With only two datasets, you can add a always-null column C to dataset 2 and use all columns.  But that may get impractical as the number of datasets and dataset-specific fields grows

There is actually a potential benefit in accuracy to the 'stacking everything' approach which is that by default it'll derive the term frequency tables from the full stacked dataset.  That could be quite useful if your individual datasets are small (and hence cardinality of the dataset is unrepresentative of the population), so that the stacked dataset is more representative.

**Comment by sjhw** [comment]
Incredible insight here! Thank you @medwar99 and @RobinL for your very thorough explanations above, as this perfectly fits with my current use case. I found great success implementing the above logic for deduplication and record linkage across 4 sources.

I'd love to see a sample notebook leveraging the above technique in the Examples tab in the docs online. I think it could be highly useful for others who want to do record linkage & deduplication across 2+ input datasets.
